### PR TITLE
fixes task edit cancellation commit

### LINKED ISSUE #7920: Checklist can't expand/collapse "Task not found"
General Info
- UUID: cf376ed3-8c46-416b-9768-68161eed8311
- Browser: Chrome
- OS: Windows 10 64bit
### Description

When attempting to expand/minimize a checklist, the item briefly turns blank, and blue, then an error appears "Task not found".  (I can edit the task, other than the due date?)  It then reverts back to the old task, with no expansion/
#### Console Errors

At top when loading page:
Deprecation warning: moment().zone is deprecated, use moment().utcOffset instead. http://momentjs.com/guides/#/warnings/zone/
Arguments: 
Error
    at s.zone (https://habitica.com/app-6f837680.js:43:12297)
    at Object.authenticate (https://habitica.com/app-6f837680.js:47:3855)
    at Object.<anonymous> (https://habitica.com/app-6f837680.js:47:5076)
    at Object.i [as invoke](https://habitica.com/app-6f837680.js:7:17125)
    at Object.c [as $get](https://habitica.com/app-6f837680.js:7:15521)
    at Object.i [as invoke](https://habitica.com/app-6f837680.js:7:17125)
    at https://habitica.com/app-6f837680.js:7:15171
    at g (https://habitica.com/app-6f837680.js:7:16758)
    at Object.i [as invoke](https://habitica.com/app-6f837680.js:7:17092)
    at D.instance (https://habitica.com/app-6f837680.js:8:8733)

When clicked:

app-6f837680.js:8 PUT https://habitica.com/api/v3/tasks/undefined?userV=9701 404 (Not Found)(anonymous function) @ app-6f837680.js:8s @ app-6f837680.js:8f @ app-6f837680.js:8j @ app-6f837680.js:9(anonymous function) @ app-6f837680.js:9$eval @ app-6f837680.js:9$digest @ app-6f837680.js:9$apply @ app-6f837680.js:9(anonymous function) @ app-6f837680.js:10dispatch @ app-6f837680.js:4r.handle @ app-6f837680.js:4
app-6f837680.js:8 XHR finished loading: PUT "https://habitica.com/api/v3/tasks/undefined?userV=9701".(anonymous function) @ app-6f837680.js:8s @ app-6f837680.js:8f @ app-6f837680.js:8j @ app-6f837680.js:9(anonymous function) @ app-6f837680.js:9$eval @ app-6f837680.js:9$digest @ app-6f837680.js:9$apply @ app-6f837680.js:9(anonymous function) @ app-6f837680.js:10dispatch @ app-6f837680.js:4r.handle @ app-6f837680.js:4
app-6f837680.js:8 XHR finished loading: GET "https://habitica.com/api/v3/user/?userV=9701".(anonymous function) @ app-6f837680.js:8s @ app-6f837680.js:8f @ app-6f837680.js:8j @ app-6f837680.js:9(anonymous function) @ app-6f837680.js:9$eval @ app-6f837680.js:9$digest @ app-6f837680.js:9$apply @ app-6f837680.js:9s @ app-6f837680.js:8v @ app-6f837680.js:8a @ app-6f837680.js:8
app-6f837680.js:8 XHR finished loading: GET "https://habitica.com/api/v3/tasks/user?userV=9701".(anonymous function) @ app-6f837680.js:8s @ app-6f837680.js:8f @ app-6f837680.js:8j @ app-6f837680.js:9(anonymous function) @ app-6f837680.js:9$eval @ app-6f837680.js:9$digest @ app-6f837680.js:9$apply @ app-6f837680.js:9s @ app-6f837680.js:8v @ app-6f837680.js:8a @ app-6f837680.js:8

### PR DESCRIPTION
**Pull request by Hus274**
Fixes #7920 closes #7920
### Changes

This reintroduces the code for fixing the task edit cancellation showing the actual task values without refreshing and fixes the issue with collapsing checklist items that was seen when that change went out.

---

UUID: 

This reverts commit 3d64f2b84971d90b2e5d46d46ca2134621bfa698.

### CODE REVIEW

File: website/views/shared/tasks/edit/advanced_options.jade [hunk 1]
```diff
@@ -1,69 +1,69 @@
-div(ng-if='::task.type!="reward"')
+div(ng-if='::task._edit.type!="reward"')
```
Comment (crookedneighbor) [inline website/views/shared/tasks/edit/advanced_options.jade]: As the type cannot change, there's no need to specify the `_edit` portion. Let's leave it as `task.type`

File: website/views/shared/tasks/edit/advanced_options.jade [hunk 2]
```diff
@@ -1,69 +1,69 @@
...
   button.advanced-options-toggle.option-title.mega(type='button',
-    ng-class='{active: task._advanced}',
-    ng-click='task._advanced = !task._advanced', tooltip=env.t('expandCollapse'))
+    ng-class='{active: task._edit._advanced}',
+    ng-click='task._edit._advanced = !task._edit._advanced', tooltip=env.t('expandCollapse'))
     =env.t('advancedOptions')
 
-  fieldset.option-group.advanced-option(ng-if="task.userId" ng-show="task._advanced")
+  fieldset.option-group.advanced-option(ng-if="task._edit.userId" ng-show="task._edit._advanced")
```
Comment (crookedneighbor) [inline website/views/shared/tasks/edit/advanced_options.jade]: Same here. The userId cannot change. No need to use the `_edit` bit.

**Comment by Hus274** [comment]
Addressed comments from @crookedneighbor

**Comment by crookedneighbor** [comment]
There are a few other places that reference the task type. It'd be great if you could remove the `_edit` bit from those too.

**Comment by Hus274** [comment]
Ahh gotcha. For sure, I'll clean this up and resubmit

**Comment by Hus274** [comment]
Removed task._edit from type

**Comment by khipkin** [comment]
@Hus274 If you add the email address you used for those commits to your GitHub profile, they'll get properly attached to your user profile.

**Comment by Hus274** [comment]
I had to reformat my hard drive and switched to working on this with my work computer which has it's own git settings for work purposes. I'm not sure I want to combine my account here with that email address. I'll give it some thought though, thanks.

File: website/views/shared/tasks/edit/habits/plus_minus.jade [hunk 3]
```diff
@@ -1,8 +1,8 @@
-fieldset.option-group.plusminus(ng-if='task.type=="habit" && canEdit(task)')
+fieldset.option-group.plusminus(ng-if='task.type === "habit" && canEdit(task)')
   legend.option-title=env.t('direction/Actions')
   span.task-checker
-    input.visuallyhidden.focusable(id='{{obj._id}}_{{task._id}}-option-plus', type='checkbox', ng-model='task.up')
-    label(for='{{obj._id}}_{{task._id}}-option-plus')
+    input.visuallyhidden.focusable(id='{{obj._id}}_{{task._edit._id}}-option-plus', type='checkbox', ng-model='task._edit.up')
```
Comment (crookedneighbor) [inline website/views/shared/tasks/edit/habits/plus_minus.jade]: All these ids can't be changed, so they should reference `task._id`, not `task._edit._id`

File: website/client/js/services/taskServices.js [hunk 4]
```diff
@@ -151,12 +151,19 @@ angular.module('habitrpg')
...
       task._advanced = !user.preferences.advancedCollapsed;
+      task._edit = angular.copy(task);
       if($rootScope.charts[task._id]) $rootScope.charts[task.id] = false;
     }
 
+    function cancelTaskEdit(task) {
+      task._edit = undefined;
+      task._editing = false;
+    };
+
```
Comment (crookedneighbor) [inline website/client/js/services/taskServices.js:188]: extra line break here.

**Comment by Hus274** [comment]
Addressed last two comments

**Comment by crookedneighbor** [comment]
Wonderful! I've notated this toward your next tier.

**Comment by SabreCat** [comment]
Sad news: the latest edition of this fix also got flaky. "Open tasks in edit mode" caused new task entries to appear without carrying forward the entered task title, then throw the blue task + "Task not found" error if the user tried to continue editing. (Tasks were correctly saved, and would show up on sync, but it was super disorienting!)

**Comment by Hus274** [comment]
Sad day. Alright, I will take another look!

**Comment by crookedneighbor** [comment]
I should have caught this when I QAd it. Working on it now.

**Comment by crookedneighbor** [comment]
Should be fixed now: https://github.com/HabitRPG/habitrpg/commit/bbe4759691b8fea39b3429fc4b52f6b903cdee94

**Comment by Hus274** [comment]
@crookedneighbor thanks for fixing it, I appreciate it. Sorry for all the issues around this! :(

**Comment by crookedneighbor** [comment]
No worries. It happens. Just glad we caught it before it was in production this time :)

**Pull request merged**